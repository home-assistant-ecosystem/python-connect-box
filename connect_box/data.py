"""Handle Data attributes."""
from ipaddress import IPv4Address, IPv6Address, ip_address as convert_ip
from typing import Union

import attr


@attr.s
class Device:
    """A single device."""

    mac: str = attr.ib()
    hostname: str = attr.ib(cmp=False)
    ip: Union[IPv4Address, IPv6Address] = attr.ib(cmp=False, converter=convert_ip)


@attr.s
class DownstreamChannel:
    """A locked downstream channel."""

    frequency: int = attr.ib()
    powerLevel: int = attr.ib()
    modulation: str = attr.ib()
    id: str = attr.ib()
    snr: float = attr.ib()
    preRs: int = attr.ib()
    postRs: int = attr.ib()
    qamLocked: bool = attr.ib()
    fecLocked: bool = attr.ib()
    mpegLocked: bool = attr.ib()


@attr.s
class UpstreamChannel:
    """A locked upstream channel."""

    frequency: int = attr.ib()
    powerLevel: int = attr.ib()
    symbolRate: str = attr.ib()
    id: str = attr.ib()
    modulation: str = attr.ib()
    type: str = attr.ib()
    t1Timeouts: int = attr.ib()
    t2Timeouts: int = attr.ib()
    t3Timeouts: int = attr.ib()
    t4Timeouts: int = attr.ib()
    channelType: str = attr.ib()
    messageType: int = attr.ib()


@attr.s
class CmStatus:
    provisioningStatus: str = attr.ib()
    cmComment: str = attr.ib()
    cmDocsisMode: str = attr.ib()
    cmNetworkAccess: str = attr.ib()
    firmwareFilename: str = attr.ib()

    # number of IP addresses to assign via DHCP
    numberOfCpes: int = attr.ib()

    # ???
    dMaxCpes: int = attr.ib()
    bpiEnable: int = attr.ib()


@attr.s
class ServiceFlow:
    id: int = attr.ib()
    pMaxTrafficRate: int = attr.ib()
    pMaxTrafficBurst: int = attr.ib()
    pMinReservedRate: int = attr.ib()
    pMaxConcatBurst: int = attr.ib()

    # 2 seems to be Best Effort
    pSchedulingType: int = attr.ib()


@attr.s
class Temperature:
    # temperatures in degrees Celsius
    tunerTemperature: float = attr.ib()
    temperature: float = attr.ib()

    # several other stats remain untapped here:
    # wan_ipv4_addr
    # wan_ipv6_addr, wan_ipv6_addr_entry
