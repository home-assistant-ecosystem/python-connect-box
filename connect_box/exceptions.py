"""Exceptions for UPC Connect Box API client."""


class ConnectBoxError(Exception):
    """General ZeroTierError exception occurred."""


class ConnectBoxConnectionError(ConnectBoxError):
    """When a connection error is encountered."""


class ConnectBoxNoDataAvailable(ConnectBoxError):
    """When no data is available."""


class ConnectBoxLoginError(ConnectBoxError):
    """When login fails."""
