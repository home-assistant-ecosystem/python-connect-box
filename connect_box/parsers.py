def _parse_general_time(xml_root):
    """Convert <GeneralTime> to"""

    el = xml_root.find("GeneralTime/time")
    if el is not None:
        return ','.join([str(int(h)*60+int(m)) for t in el.text.split('-') for h, m in [t.split(':')]])
    else:
        return None

def _parse_daily_time(xml_root):
    """Convert contents of <DailyTime> to decimal mask"""
    mask = [[0 for h in range(24)] for d in range(7)]
    for instance in xml_root.findall("DailyTime/time_instance"):
        day = int(instance.find('daily').text) -1
        s, e = instance.find('time').text.split('-')
        s, e = int(s), int(e)+1
        mask[day][s:e] = [1] * (e-s)

    output = []
    for daybin in mask:
        daydec = 0
        for bit in daybin:
            daydec = (daydec  << 1) | bit
        output.append(str(daydec))

    if (output is not None and len(output)):
        return ','.join(output)
    else:
        return None
