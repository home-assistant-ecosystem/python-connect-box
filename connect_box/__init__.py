"""A Python Client to get data from UPC Connect Boxes."""
import asyncio
import logging
from typing import Dict, List, Optional

import aiohttp
from aiohttp.hdrs import REFERER, USER_AGENT
import defusedxml.ElementTree as element_tree

from .data import (
    Device,
    DownstreamChannel,
    UpstreamChannel,
    CmStatus,
    ServiceFlow,
    Temperature,
)
from . import exceptions

_LOGGER = logging.getLogger(__name__)

HTTP_HEADER_X_REQUESTED_WITH = "X-Requested-With"

CMD_LOGIN = 15
CMD_LOGOUT = 16
CMD_DEVICES = 123
CMD_DOWNSTREAM = 10
CMD_UPSTREAM = 11
CMD_TEMPERATURE = 136
CMD_CMSTATUS = 144


class ConnectBox:
    """A class for handling the data retrieval from an UPC Connect Box."""

    def __init__(
        self, session: aiohttp.ClientSession, password: str, host: str = "192.168.0.1"
    ):
        """Initialize the connection."""
        self._session: aiohttp.ClientSession = session
        self.token: Optional[str] = None
        self.host: str = host
        self.password: str = password
        self.headers: Dict[str, str] = {
            HTTP_HEADER_X_REQUESTED_WITH: "XMLHttpRequest",
            REFERER: f"http://{self.host}/index.html",
            USER_AGENT: (
                "Mozilla/5.0 (Windows NT 10.0; WOW64) "
                "AppleWebKit/537.36 (KHTML, like Gecko) "
                "Chrome/47.0.2526.106 Safari/537.36"
            ),
        }
        self.devices: List[Device] = []
        self.ds_channels: List[DownstreamChannel] = []
        self.us_channels: List[UpstreamChannel] = []
        self.cmstatus: Optional[CmStatus] = None
        self.upstream_service_flows: List[ServiceFlow] = []
        self.downstream_service_flows: List[ServiceFlow] = []
        self.temperature: Optional[Temperature] = None

    async def async_get_devices(self):
        """Scan for new devices and return a list with found device IDs."""
        if self.token is None:
            await self.async_initialize_token()

        self.devices.clear()
        raw = await self._async_ws_function(CMD_DEVICES)

        try:
            xml_root = element_tree.fromstring(raw)
            mac_adresses: List[str] = [mac.text for mac in xml_root.iter("MACAddr")]
            hostnames: List[str] = [mac.text for mac in xml_root.iter("hostname")]
            ip_addresses: List[str] = [mac.text for mac in xml_root.iter("IPv4Addr")]

            for mac_address, hostname, ip_address in zip(
                mac_adresses, hostnames, ip_addresses
            ):
                self.devices.append(
                    Device(mac_address, hostname, ip_address.partition("/")[0])
                )
        except (element_tree.ParseError, TypeError):
            _LOGGER.warning("Can't read device from %s", self.host)
            self.token = None
            raise exceptions.ConnectBoxNoDataAvailable() from None

    async def async_get_downstream(self):
        """Get the current downstream cable modem state."""
        if self.token is None:
            await self.async_initialize_token()

        self.ds_channels.clear()
        raw = await self._async_ws_function(CMD_DOWNSTREAM)

        try:
            xml_root = element_tree.fromstring(raw)
            for downstream in xml_root.iter("downstream"):
                self.ds_channels.append(
                    DownstreamChannel(
                        int(downstream.find("freq").text),
                        int(downstream.find("pow").text),
                        downstream.find("mod").text,
                        downstream.find("chid").text,
                        float(downstream.find("RxMER").text),
                        int(downstream.find("PreRs").text),
                        int(downstream.find("PostRs").text),
                        downstream.find("IsQamLocked").text == "1",
                        downstream.find("IsFECLocked").text == "1",
                        downstream.find("IsMpegLocked").text == "1",
                    )
                )
        except (element_tree.ParseError, TypeError):
            _LOGGER.warning("Can't read downstream channels from %s", self.host)
            self.token = None
            raise exceptions.ConnectBoxNoDataAvailable() from None

    async def async_get_upstream(self):
        """Get the current upstream cable modem state."""
        if self.token is None:
            await self.async_initialize_token()

        self.us_channels.clear()
        raw = await self._async_ws_function(CMD_UPSTREAM)

        try:
            xml_root = element_tree.fromstring(raw)
            for upstream in xml_root.iter("upstream"):
                self.us_channels.append(
                    UpstreamChannel(
                        int(upstream.find("freq").text),
                        int(upstream.find("power").text),
                        upstream.find("srate").text,
                        upstream.find("usid").text,
                        upstream.find("mod").text,
                        upstream.find("ustype").text,
                        int(upstream.find("t1Timeouts").text),
                        int(upstream.find("t2Timeouts").text),
                        int(upstream.find("t3Timeouts").text),
                        int(upstream.find("t4Timeouts").text),
                        upstream.find("channeltype").text,
                        int(upstream.find("messageType").text),
                    )
                )
        except (element_tree.ParseError, TypeError):
            _LOGGER.warning("Can't read upstream channels from %s", self.host)
            self.token = None
            raise exceptions.ConnectBoxNoDataAvailable() from None

    async def async_get_cmstatus_and_service_flows(self):
        """Get various status information."""
        if self.token is None:
            await self.async_initialize_token()

        self.cmstatus = None
        self.downstream_service_flows = []
        self.upstream_service_flows = []
        raw = await self._async_ws_function(CMD_CMSTATUS)

        try:
            xml_root = element_tree.fromstring(raw)
            self.cmstatus = CmStatus(
                provisioningStatus=xml_root.find("provisioning_st").text,
                cmComment=xml_root.find("cm_comment").text,
                cmDocsisMode=xml_root.find("cm_docsis_mode").text,
                cmNetworkAccess=xml_root.find("cm_network_access").text,
                numberOfCpes=int(xml_root.find("NumberOfCpes").text),
                firmwareFilename=xml_root.find("FileName").text,
                dMaxCpes=int(xml_root.find("dMaxCpes").text),
                bpiEnable=int(xml_root.find("bpiEnable").text),
            )
            for elmt_service_flow in xml_root.iter("serviceflow"):
                service_flow = ServiceFlow(
                    id=int(elmt_service_flow.find("Sfid").text),
                    pMaxTrafficRate=int(elmt_service_flow.find("pMaxTrafficRate").text),
                    pMaxTrafficBurst=int(
                        elmt_service_flow.find("pMaxTrafficBurst").text
                    ),
                    pMinReservedRate=int(
                        elmt_service_flow.find("pMinReservedRate").text
                    ),
                    pMaxConcatBurst=int(elmt_service_flow.find("pMaxConcatBurst").text),
                    pSchedulingType=int(elmt_service_flow.find("pSchedulingType").text),
                )
                direction = int(elmt_service_flow.find("direction").text)
                if direction == 1:
                    self.downstream_service_flows.append(service_flow)
                elif direction == 2:
                    self.upstream_service_flows.append(service_flow)
                else:
                    raise element_tree.ParseError(
                        "Unknown service flow direction '{}'".format(direction)
                    )
        except (element_tree.ParseError, TypeError):
            _LOGGER.warning("Can't read cmstatus from %s", self.host)
            self.token = None
            raise exceptions.ConnectBoxNoDataAvailable() from None

    async def async_get_temperature(self):
        """Get temperature information (in degrees Celsius)."""
        if self.token is None:
            await self.async_initialize_token()

        self.temperature = None
        raw = await self._async_ws_function(CMD_TEMPERATURE)

        f_to_c = lambda f: (5.0 / 9) * (f - 32)
        try:
            xml_root = element_tree.fromstring(raw)
            self.temperature = Temperature(
                tunerTemperature=f_to_c(int(xml_root.find("TunnerTemperature").text)),
                temperature=f_to_c(int(xml_root.find("Temperature").text)),
            )
        except (element_tree.ParseError, TypeError):
            _LOGGER.warning("Can't read temperature from %s", self.host)
            self.token = None
            raise exceptions.ConnectBoxNoDataAvailable() from None

    async def async_close_session(self) -> None:
        """Logout and close session."""
        if not self.token:
            return

        await self._async_ws_function(CMD_LOGOUT)
        self.token = None

    async def async_initialize_token(self) -> None:
        """Get the token first."""
        try:
            # Get first the token
            async with self._session.get(
                f"http://{self.host}/common_page/login.html",
                headers=self.headers,
                timeout=10,
            ) as response:
                await response.text()
                self.token = response.cookies["sessionToken"].value

        except (asyncio.TimeoutError, aiohttp.ClientError) as err:
            _LOGGER.error("Can not load login page from %s: %s", self.host, err)
            raise exceptions.ConnectBoxConnectionError()

        await self._async_initialize_token_with_password(CMD_LOGIN)

    async def _async_initialize_token_with_password(self, function: int) -> None:
        """Get token with password."""
        try:
            async with await self._session.post(
                f"http://{self.host}/xml/setter.xml",
                data=f"token={self.token}&fun={function}&Username=NULL&Password={self.password}",
                headers=self.headers,
                allow_redirects=False,
                timeout=10,
            ) as response:
                await response.text()

                if response.status != 200:
                    _LOGGER.warning("Login error with code %d", response.status)
                    self.token = None
                    raise exceptions.ConnectBoxLoginError()
                self.token = response.cookies["sessionToken"].value

        except (asyncio.TimeoutError, aiohttp.ClientError) as err:
            _LOGGER.error("Can not login to %s: %s", self.host, err)
            raise exceptions.ConnectBoxConnectionError()

    async def _async_ws_function(self, function: int) -> Optional[str]:
        """Execute a command on UPC firmware webservice."""
        try:
            # The 'token' parameter has to be first, and 'fun' second
            # or the UPC firmware will return an error
            async with await self._session.post(
                f"http://{self.host}/xml/getter.xml",
                data=f"token={self.token}&fun={function}",
                headers=self.headers,
                allow_redirects=False,
                timeout=10,
            ) as response:

                # If there is an error
                if response.status != 200:
                    _LOGGER.debug("Receive HTTP code %d", response.status)
                    self.token = None
                    raise exceptions.ConnectBoxError()

                # Load data, store token for next request
                self.token = response.cookies["sessionToken"].value
                return await response.text()

        except (asyncio.TimeoutError, aiohttp.ClientError) as err:
            _LOGGER.error("Error received on %s: %s", function, err)
            self.token = None

        raise exceptions.ConnectBoxConnectionError()
