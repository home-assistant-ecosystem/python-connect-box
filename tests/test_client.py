from contextlib import contextmanager
import aiohttp

import pytest
from connect_box import ConnectBox


@pytest.mark.asyncio
async def test_async_initialize_token(client, vcr_recorder):
    with vcr_recorder():
        await client.async_initialize_token()

    assert client.token == "[SESSION_TOKEN]"


@pytest.mark.asyncio
async def test_async_get_devices(client, vcr_recorder, snapshot):
    with vcr_recorder():
        await client.async_get_devices()

    assert client.devices == snapshot


@pytest.mark.asyncio
async def test_async_get_downstream(client, vcr_recorder, snapshot):
    with vcr_recorder():
        await client.async_get_downstream()

    assert client.ds_channels == snapshot


@pytest.mark.asyncio
async def test_async_get_devices(client, vcr_recorder, snapshot):
    with vcr_recorder():
        await client.async_get_devices()

    assert client.devices == snapshot


@pytest.mark.asyncio
async def test_async_get_ipv6_filtering(client, vcr_recorder, snapshot):
    with vcr_recorder():
        await client.async_get_ipv6_filtering()

    assert client.ipv6_filters == snapshot


@pytest.mark.asyncio
async def test_async_async_toggle_ipv6_filter(client, vcr_recorder, snapshot):
    with vcr_recorder():
        result = await client.async_toggle_ipv6_filter(1)

    assert result == snapshot


@pytest.mark.asyncio
async def test_async_get_lanstatus(client, vcr_recorder, snapshot):
    with vcr_recorder():
        await client.async_get_lanstatus()

    assert client.lanstatus == snapshot


@pytest.mark.asyncio
async def test_async_get_wanstatus(client, vcr_recorder, snapshot):
    with vcr_recorder():
        await client.async_get_wanstatus()

    assert client.wanstatus == snapshot


@pytest.mark.asyncio
async def test_async_get_wanstatus(client, vcr_recorder, snapshot):
    with vcr_recorder():
        await client.async_get_cm_system_info()

    assert client.cm_systeminfo == snapshot


@pytest.mark.asyncio
async def test_async_get_wanstatus(client, vcr_recorder, snapshot):
    with vcr_recorder():
        await client.async_get_cmstatus_and_service_flows()

    assert client.cmstatus == snapshot
    assert client.downstream_service_flows == snapshot
    assert client.upstream_service_flows == snapshot


@pytest.mark.asyncio
async def test_async_get_temperature(client, vcr_recorder, snapshot):
    with vcr_recorder():
        await client.async_get_temperature()

    assert client.temperature == snapshot


@pytest.mark.asyncio
async def test_async_get_eventlog(client, vcr_recorder, snapshot):
    with vcr_recorder():
        await client.async_get_eventlog()

    assert client.eventlog == snapshot


@pytest.mark.asyncio
async def test_async_reboot_device(client, vcr_recorder, snapshot):
    with vcr_recorder():
        await client.async_reboot_device()

    assert client.token == None


@pytest.mark.asyncio
async def test_async_close_session(client, vcr_recorder):
    with vcr_recorder():
        await client.async_close_session()


@pytest.mark.asyncio
async def test_async_get_global_settings(client, vcr_recorder, snapshot):
    with vcr_recorder():
        await client.async_get_global_settings()

    assert client.global_settings == snapshot


@pytest.mark.asyncio
async def test_async_get_global_settings(client, vcr_recorder, snapshot):
    with vcr_recorder():
        await client.async_get_global_settings()

    assert client.global_settings == snapshot
