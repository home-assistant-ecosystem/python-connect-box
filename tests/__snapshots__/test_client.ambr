# serializer version: 1
# name: test_async_async_toggle_ipv6_filter
  None
# ---
# name: test_async_get_devices
  list([
    Device(mac='00:01:02:03:04:05', hostname='unknown', ip=IPv4Address('192.168.0.31'), interface='Other', speed='1000', interface_id='1', method='1', lease_time=datetime.datetime(1900, 1, 1, 23, 56, 46)),
    Device(mac='00:02:04:06:08:0A', hostname='MBP', ip=IPv4Address('192.168.0.129'), interface='UPC', speed='585', interface_id='19', method='1', lease_time=datetime.datetime(1900, 1, 1, 23, 59, 59)),
  ])
# ---
# name: test_async_get_downstream
  list([
    DownstreamChannel(frequency=794000000, powerLevel=-3, modulation='256qam', id='24', snr=34.484, preRs=65471343180, postRs=6143480, qamLocked=True, fecLocked=True, mpegLocked=True),
    DownstreamChannel(frequency=786000000, powerLevel=-5, modulation='256qam', id='23', snr=34.484, preRs=65476818580, postRs=670720, qamLocked=True, fecLocked=True, mpegLocked=True),
    DownstreamChannel(frequency=778000000, powerLevel=-4, modulation='256qam', id='22', snr=34.926, preRs=65477054174, postRs=427162, qamLocked=True, fecLocked=True, mpegLocked=True),
    DownstreamChannel(frequency=770000000, powerLevel=-3, modulation='256qam', id='21', snr=35.595, preRs=65477255647, postRs=232536, qamLocked=True, fecLocked=True, mpegLocked=True),
    DownstreamChannel(frequency=762000000, powerLevel=-3, modulation='256qam', id='20', snr=35.78, preRs=65477318701, postRs=172105, qamLocked=True, fecLocked=True, mpegLocked=True),
    DownstreamChannel(frequency=754000000, powerLevel=-4, modulation='256qam', id='19', snr=34.926, preRs=65477045764, postRs=438973, qamLocked=True, fecLocked=True, mpegLocked=True),
    DownstreamChannel(frequency=746000000, powerLevel=-4, modulation='256qam', id='18', snr=34.346, preRs=65476479492, postRs=1011196, qamLocked=True, fecLocked=True, mpegLocked=True),
    DownstreamChannel(frequency=738000000, powerLevel=-3, modulation='256qam', id='17', snr=35.595, preRs=65477333381, postRs=158466, qamLocked=True, fecLocked=True, mpegLocked=True),
    DownstreamChannel(frequency=730000000, powerLevel=-3, modulation='256qam', id='16', snr=35.78, preRs=65477340289, postRs=144149, qamLocked=True, fecLocked=True, mpegLocked=True),
    DownstreamChannel(frequency=722000000, powerLevel=-2, modulation='256qam', id='15', snr=36.61, preRs=65477440415, postRs=45309, qamLocked=True, fecLocked=True, mpegLocked=True),
    DownstreamChannel(frequency=714000000, powerLevel=-2, modulation='256qam', id='14', snr=34.926, preRs=65476443083, postRs=1049850, qamLocked=True, fecLocked=True, mpegLocked=True),
    DownstreamChannel(frequency=706000000, powerLevel=-2, modulation='256qam', id='13', snr=33.487, preRs=65470830487, postRs=6658966, qamLocked=True, fecLocked=True, mpegLocked=True),
    DownstreamChannel(frequency=698000000, powerLevel=-2, modulation='256qam', id='12', snr=34.346, preRs=65477102100, postRs=391515, qamLocked=True, fecLocked=True, mpegLocked=True),
    DownstreamChannel(frequency=690000000, powerLevel=0, modulation='256qam', id='11', snr=34.346, preRs=65447317243, postRs=30138988, qamLocked=True, fecLocked=True, mpegLocked=True),
    DownstreamChannel(frequency=682000000, powerLevel=0, modulation='256qam', id='10', snr=36.387, preRs=65477467552, postRs=30148, qamLocked=True, fecLocked=True, mpegLocked=True),
    DownstreamChannel(frequency=674000000, powerLevel=-1, modulation='256qam', id='9', snr=36.387, preRs=65477459255, postRs=31177, qamLocked=True, fecLocked=True, mpegLocked=True),
    DownstreamChannel(frequency=666000000, powerLevel=0, modulation='256qam', id='8', snr=36.61, preRs=65477470925, postRs=25136, qamLocked=True, fecLocked=True, mpegLocked=True),
    DownstreamChannel(frequency=658000000, powerLevel=0, modulation='256qam', id='7', snr=36.61, preRs=65477468243, postRs=27236, qamLocked=True, fecLocked=True, mpegLocked=True),
    DownstreamChannel(frequency=650000000, powerLevel=0, modulation='256qam', id='6', snr=37.356, preRs=65477472275, postRs=24150, qamLocked=True, fecLocked=True, mpegLocked=True),
    DownstreamChannel(frequency=642000000, powerLevel=2, modulation='256qam', id='5', snr=37.356, preRs=65477475289, postRs=21606, qamLocked=True, fecLocked=True, mpegLocked=True),
    DownstreamChannel(frequency=634000000, powerLevel=1, modulation='256qam', id='4', snr=37.356, preRs=65477476542, postRs=20832, qamLocked=True, fecLocked=True, mpegLocked=True),
    DownstreamChannel(frequency=626000000, powerLevel=2, modulation='256qam', id='3', snr=37.356, preRs=65477477890, postRs=19152, qamLocked=True, fecLocked=True, mpegLocked=True),
    DownstreamChannel(frequency=610000000, powerLevel=2, modulation='256qam', id='1', snr=37.636, preRs=65477481554, postRs=16825, qamLocked=True, fecLocked=True, mpegLocked=True),
    DownstreamChannel(frequency=618000000, powerLevel=2, modulation='256qam', id='2', snr=37.636, preRs=65478719484, postRs=17895, qamLocked=True, fecLocked=True, mpegLocked=True),
  ])
# ---
# name: test_async_get_eventlog
  list([
    LogEvent(evPrio='notice', evMsg='Illegal\n        -  Dropped FORWARD packet: SRC=2a03:2880:f216:00c6:face:b00c:0000:7260 MAC=56:2B:76:F0:19:86;CM-MAC=38:43:7d:d1:ff:ff;CMTS-MAC=00:56:2b:76:ff:ff;CM-QOS=1.1;CM-VER=3.0;', evTime='03-01-2024\n        19:25:43', evEpoch=1704306343),
  ])
# ---
# name: test_async_get_global_settings
  GlobalSettings(logged_in=True, operator_id='[OPERATOR]', access_denied=False, sw_version=None)
# ---
# name: test_async_get_ipv6_filtering
  list([
  ])
# ---
# name: test_async_get_lanstatus
  LanStatus(upnp_enabled=True, mac='38:43:7D:FF:FF:FF', ip4=IPv4Address('192.168.0.1'), ip6=IPv6Network('2a02:a317:2244:3400::/64'))
# ---
# name: test_async_get_temperature
  Temperature(tunerTemperature=31.111111111111114, temperature=20.0)
# ---
# name: test_async_get_wanstatus
  CmStatus(provisioningStatus='Online', cmComment='Operational', cmDocsisMode='DOCSIS\n        3.0', cmNetworkAccess='Allowed', firmwareFilename='CH7465LG_fiberpower008_v6.cfg', numberOfCpes=245, dMaxCpes=2, bpiEnable=1)
# ---
# name: test_async_get_wanstatus.1
  list([
    ServiceFlow(id=324, pMaxTrafficRate=768000000, pMaxTrafficBurst=42600, pMinReservedRate=0, pMaxConcatBurst=1522, pSchedulingType=1),
    ServiceFlow(id=23569, pMaxTrafficRate=53000000, pMaxTrafficBurst=16320, pMinReservedRate=0, pMaxConcatBurst=1522, pSchedulingType=1),
  ])
# ---
# name: test_async_get_wanstatus.2
  list([
    ServiceFlow(id=323, pMaxTrafficRate=40960000, pMaxTrafficBurst=42600, pMinReservedRate=0, pMaxConcatBurst=42600, pSchedulingType=2),
    ServiceFlow(id=23568, pMaxTrafficRate=5300000, pMaxTrafficBurst=16320, pMinReservedRate=0, pMaxConcatBurst=16320, pSchedulingType=2),
  ])
# ---
